### PR TITLE
delete image component.

### PR DESCRIPTION
**Pull request by akash1810**
Reusing `gr-confirm-delete` to show a confirm message before performing a delete on an image.

In first instance, adding this to the preview page:

delete-image

cc @alastairjardine

### CODE REVIEW

File: kahuna/public/js/edits/image-editor.js [hunk 1]
```diff
@@ -73,6 +75,18 @@ imageEditor.controller('ImageEditorCtrl',
         ctrl.error = true;
     }
 
+    mediaApi.canDelete(ctrl.image).then(deletable => {
+        ctrl.canBeDeleted = deletable;
+    });
+
+
+    ctrl.onDeleteSuccess = () => {
+        $state.go('upload', {}, {reload: true});
```
Comment (theefer) [inline kahuna/public/js/edits/image-editor.js]: Does this re-load the whole page?

File: kahuna/public/js/components/gr-delete-image/gr-delete-image.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+export const imageDelete = angular.module('gr.imageDelete', ['gr.confirmDelete']);
+
+imageDelete.controller('grImageDeleteCtrl', ['$scope', 'mediaApi', function ($scope, mediaApi) {
+    var ctrl = this;
+
+    ctrl.delete = function () {
+        mediaApi.delete(ctrl.image)
+            .then((resp) => ctrl.onSuccess(resp))
+            .catch((err) => ctrl.onError(err));
```
Comment (theefer) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: Note that if either of these callbacks are not provided this will error out. If they are required, could check for their presence when initialising the directive (fail early)?

File: kahuna/public/js/edits/image-editor.js (hunk 1 above)
Comment (akash1810) [inline kahuna/public/js/edits/image-editor.js]: Yeah :cry:

Ideally just that `li` would be removed from the page. Will take a look.

**Comment by alastairjardine** [comment]
Cool – could we update the copy so its:

Delete —> Confirm delete.

Thanks

File: kahuna/public/js/components/gr-confirm-delete/gr-confirm-delete.js [hunk 3]
```diff
@@ -6,10 +7,14 @@ confirmDelete.directive('grConfirmDelete', ['$timeout', function($timeout) {
...
         template: `
-            <button class="button-ico" type="button" ng:click="showConfirm = true">
-                <span class="confirm" ng:if="showConfirm">Sure?</span>
-                <gr-icon>undo</gr-icon>
+            <button class="gr-confirm-delete button-ico" type="button"
+                ng:click="showConfirm = true"
+                ng:class="{'gr-confirm-delete--confirm': showConfirm}">
+                <gr-icon>delete</gr-icon>
+                <ng-transclude ng:if="!showConfirm"></ng-transclude>
+                <span class="confirm" ng:if="showConfirm">are you sure?</span>
```
Comment (alastairjardine) [inline kahuna/public/js/components/gr-confirm-delete/gr-confirm-delete.js]: "Confirm delete"

Comment (jamesgorrie) [inline kahuna/public/js/components/gr-confirm-delete/gr-confirm-delete.js]: > "Confirm delete"

:computer::lips:

File: kahuna/public/js/components/gr-delete-image/gr-delete-image.js (hunk 2 above)
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: Would this not error from instantiating the directive without them as you've required them in the [scope](https://github.com/guardian/grid/pull/1051/files#diff-98f40c89c6f7ce3607c23deb9b0b1856R29)?

Comment (jamesgorrie) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: You would use '=?' if you wanted them as optional.

File: kahuna/public/js/image/controller.js [hunk 4]
```diff
@@ -176,16 +178,15 @@ image.controller('ImageCtrl', [
...
-                });
+        ctrl.onDeleteSuccess = function () {
+            $state.go('search');
+        };
+
+        ctrl.onDeleteError = function (err) {
+            if (err.body.errorKey === 'image-not-found') {
+                $state.go('search');
+            } else {
+                window.alert(err.body.errorMessage);
```
Comment (jamesgorrie) [inline kahuna/public/js/image/controller.js]: `$window`?

File: kahuna/public/js/edits/image-editor.js [hunk 5]
```diff
@@ -73,6 +75,18 @@ imageEditor.controller('ImageEditorCtrl',
...
+    };
+
+    ctrl.onDeleteError = (err) => {
+        window.alert(err.body.errorMessage);
```
Comment (jamesgorrie) [inline kahuna/public/js/edits/image-editor.js]: `$window`?

**Comment by jamesgorrie** [comment]
There is the question on the uploads page as to whether this acts more like a "cancel" than a delete, not sure we should just bodge it in there?

File: kahuna/public/js/components/gr-delete-image/gr-delete-image.js (hunk 2 above)
Comment (theefer) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: Oh really? I thought they'd end up undefined. If that errors automatically then great.

**Comment by theefer** [comment]
"cancel" would perpetuate the false myth that images aren't uploaded yet. They are. I think we should make it clearer.

**Comment by jamesgorrie** [comment]
> They are. I think we should make it clearer.

:+1: +

File: kahuna/public/js/components/gr-delete-image/gr-delete-image.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+            })
+            .catch((err) => {
+                if (angular.isDefined(ctrl.onError)) {
+                    ctrl.onError(err);
+                }
+            });
+    };
+}]);
+
+imageDelete.directive('grImageDelete', [function () {
```
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: Is there a reason why the directive / ctrl and file name are different - cold be confusing when searching.

File: kahuna/public/js/edits/image-editor.html [hunk 7]
```diff
@@ -75,6 +75,14 @@
             ></ui-archiver>
         </div>
 
+        <div class="result-editor__delete-image">
+            <gr-image-delete image="ctrl.image"
```
Comment (jamesgorrie) [inline kahuna/public/js/edits/image-editor.html]: We should conventionalise on naming i.e. `-` vs `:`

Comment (jamesgorrie) [inline kahuna/public/js/edits/image-editor.html]: ^ Again - making it hard to search the codebase

Comment (theefer) [inline kahuna/public/js/edits/image-editor.html]: I think we always use `-` in element names for browser compat/styling/etc reasons?

File: kahuna/public/js/components/gr-delete-image/gr-delete-image.js (hunk 6 above)
Comment (akash1810) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: I didn't even notice that :fearful:

File: kahuna/public/js/edits/image-editor.html (hunk 7 above)
Comment (jamesgorrie) [inline kahuna/public/js/edits/image-editor.html]: ope, not what I've been doing, I'll go clean it up.

File: kahuna/public/js/components/gr-confirm-delete/gr-confirm-delete.js (hunk 3 above)
Comment (akash1810) [inline kahuna/public/js/components/gr-confirm-delete/gr-confirm-delete.js]: I might as well update all the items in the toolbar in this PR too - it looks weird otherwise.

Comment (jamesgorrie) [inline kahuna/public/js/components/gr-confirm-delete/gr-confirm-delete.js]: clean up as we go along is a good way to do it.

Comment (theefer) [inline kahuna/public/js/components/gr-confirm-delete/gr-confirm-delete.js]: Just a note: let's keep separate functional changes in separate PRs, that makes it easier to review, revert, avoids blocking individual functionalities when discussing unrelated changes, etc. Thanks.

File: kahuna/public/js/edits/image-editor.js (hunk 5 above)
Comment (akash1810) [inline kahuna/public/js/edits/image-editor.js]: Yep.

Will make it talk to the error service in a later PR too.

File: kahuna/public/js/edits/image-editor.js (hunk 1 above)
Comment (theefer) [inline kahuna/public/js/edits/image-editor.js]: Suggested emitting a notification, similar to image-updated, which the image upload service could listen to to update its state (remove deleted images).

Comment (jamesgorrie) [inline kahuna/public/js/edits/image-editor.js]: After @kenoir  and I worked on the image result list with this - it was confusing.
Hence the push to move to the [stream thing](https://github.com/guardian/grid/pull/1022).

As we can explicitly tie the list and edits, I think we should do it that way (have the image list listen to edits somehow, not on a global handler).

Comment (akash1810) [inline kahuna/public/js/edits/image-editor.js]: In this instance, I don't think a global event emission is required. Instead, the `gr-delete-image` shouldn't actually belong in `ui-image-editor` instead, it should be [here](https://github.com/guardian/grid/blob/master/kahuna/public/js/upload/view.html#L28). That way, the upload controller can remove the image from `this.myUploads` in the `onSuccess` argument of `gr-delete-image`.

Comment (jamesgorrie) [inline kahuna/public/js/edits/image-editor.js]: @akash1810 :+1:

**Comment by kenoir** [comment]
@akash1810 do you reckon we should keep the button a consistent size so when the text changes the width remains the same?

**Comment by akash1810** [comment]
@kenoir definitely!

**Comment by jamesgorrie** [comment]
> a consistent size

Perhaps setting a `min-width`, just incase it outgrows itself.

**Comment by akash1810** [comment]
Updated this PR to only change the Preview page - will add to Uploads page in a subsequent PR.

File: kahuna/public/js/components/gr-delete-image/gr-delete-image.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import angular from 'angular';
+import '../gr-confirm-delete/gr-confirm-delete';
+
+export const deleteImage = angular.module('gr.deleteImage', ['gr.confirmDelete']);
+
+deleteImage.controller('grDeleteImageCtrl', ['$rootScope', '$scope', 'mediaApi',
+    function ($rootScope, $scope, mediaApi) {
```
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: `$rootScope` <= redundant?

Comment (jamesgorrie) [inline kahuna/public/js/components/gr-delete-image/gr-delete-image.js]: `$scope` too?

File: kahuna/public/js/edits/image-editor.js [hunk 9]
```diff
@@ -9,12 +9,12 @@ import '../usage-rights/usage-rights-editor';
 export var imageEditor = angular.module('kahuna.edits.imageEditor', [
     'kahuna.edits.service',
     'gr.image.service',
-    'kahuna.edits.usageRightsEditor'
+    'kahuna.edits.usageRightsEditor',
```
Comment (jamesgorrie) [inline kahuna/public/js/edits/image-editor.js]: Trailing commas at end of array can cause problems...

**Comment by jamesgorrie** [comment]
:+1:
